### PR TITLE
(bugfix): improve `gh-r` binary selection logic

### PR DESCRIPTION
**Pull request by vladdoster**
- fix 🐞 bug: `bpick` selects Darwin binary when using Manjaro (Arch
  Linux) #160
- fix edge cases where Zinit selects a Linux binary on MacOS and
  vice-versa
- fix zinit not selecting musl version on musl systems
- add unit tests for `gh-r` on Ubuntu & MacOS systems to verify logic

Signed-off-by: Vladislav Doster <mvdoster@gmail.com>

### CODE REVIEW

File: zinit-install.zsh [hunk 1]
```diff
@@ -1503,22 +1516,31 @@ builtin source "${ZINIT[BIN_DIR]}/zinit-side.zsh" || {
...
+        if (( ${#list} > 1 && ${#matchstr[${MACHTYPE}-2]} )) {
+            list2=( ${(M)list[@]:#(#i)*${~matchstr[${MACHTYPE}-2]:-${MACHTYPE#(#i)(i|amd)}}*} )
             (( $#list2 > 0 )) && list=( ${list2[@]} )
         }
 
         if (( $#list > 1 )) {
-            list2=( ${list[@]:#(#i)*.(sha[[:digit:]]#|asc)} )
+            list2=( ${(M)list[@]:#(#i)*${~matchstr[$OSTYPE(#i)]:-${OSTYPE#(#i)}}*} )
+
+            +zinit-message "{pre}.zinit-get-latest-gh-r-url-part{msg2} DEBUG:{msg}" \
```
Comment (vladdoster) [inline zinit-install.zsh:1531]: Added as debug logging in case someone encounters edge case while testing.

Will be removed.

File: zinit-install.zsh [hunk 2]
```diff
@@ -1447,25 +1447,25 @@ builtin source "${ZINIT[BIN_DIR]}/zinit-side.zsh" || {
         local url=https://$urlpart
     }
 
+        #  linux   "((#a4)linuxmusl|musl|linux(-|_)gnu]|gnu)"
```
Comment (vladdoster) [inline zinit-install.zsh:1450]: Will be removed.

**Comment by sm1999** [comment]
I tested this pr and found 2 edge cases for my system( linux amd64 gnu )

1. For ripgrep `ripgrep-13.0.0-arm-unknown-linux-gnueabihf.tar.gz` release was selected.

<details>
<summary>DEBUG</summary>

```.zinit-get-latest-gh-r-url-part DEBUG: `/BurntSushi/ripgrep/releases/download/13.0.0/ripgrep-13.0.0-arm-unknown-linux-gnueabihf.tar.gz /BurntSushi/ripgrep/releases/download/12.1.1/ripgrep-12.1.1-arm-unknown-linux-gnueabihf.tar.gz /BurntSushi/ripgrep/releases/download/12.1.0/ripgrep-12.1.0-arm-unknown-linux-gnueabihf.tar.gz /BurntSushi/ripgrep/releases/download/12.0.1/ripgrep-12.0.1-arm-unknown-linux-gnueabihf.tar.gz /BurntSushi/ripgrep/releases/download/12.0.0/ripgrep-12.0.0-arm-unknown-linux-gnueabihf.tar.gz /BurntSushi/ripgrep/releases/download/11.0.2/ripgrep-11.0.2-arm-unknown-linux-gnueabihf.tar.gz /BurntSushi/ripgrep/releases/download/11.0.1/ripgrep-11.0.1-arm-unknown-linux-gnueabihf.tar.gz /BurntSushi/ripgrep/releases/download/11.0.0/ripgrep-11.0.0-arm-unknown-linux-gnueabihf.tar.gz /BurntSushi/ripgrep/releases/download/0.10.0/ripgrep-0.10.0-arm-unknown-linux-gnueabihf.tar.gz /BurntSushi/ripgrep/releases/download/0.9.0/ripgrep-0.9.0-arm-unknown-linux-gnueabihf.tar.gz'```

</details>

2. For exa `exa-linux-x86_64-musl-v0.10.1.zip` release was selected over `exa-linux-x86_64-v0.10.1.zip`
(might be because of no release containing gnu)

<details>
<summary>DEBUG</summary>

.zinit-get-latest-gh-r-url-part DEBUG: `'

</details>

**Comment by vladdoster** [comment]
> I tested this pr and found 2 edge cases for my system( linux amd64 gnu )
> 
> 1. For ripgrep `ripgrep-13.0.0-arm-unknown-linux-gnueabihf.tar.gz` release was selected.
> 
> DEBUG
> `` .zinit-get-latest-gh-r-url-part DEBUG: `/BurntSushi/ripgrep/releases/download/13.0.0/ripgrep-13.0.0-arm-unknown-linux-gnueabihf.tar.gz /BurntSushi/ripgrep/releases/download/12.1.1/ripgrep-12.1.1-arm-unknown-linux-gnueabihf.tar.gz /BurntSushi/ripgrep/releases/download/12.1.0/ripgrep-12.1.0-arm-unknown-linux-gnueabihf.tar.gz /BurntSushi/ripgrep/releases/download/12.0.1/ripgrep-12.0.1-arm-unknown-linux-gnueabihf.tar.gz /BurntSushi/ripgrep/releases/download/12.0.0/ripgrep-12.0.0-arm-unknown-linux-gnueabihf.tar.gz /BurntSushi/ripgrep/releases/download/11.0.2/ripgrep-11.0.2-arm-unknown-linux-gnueabihf.tar.gz /BurntSushi/ripgrep/releases/download/11.0.1/ripgrep-11.0.1-arm-unknown-linux-gnueabihf.tar.gz /BurntSushi/ripgrep/releases/download/11.0.0/ripgrep-11.0.0-arm-unknown-linux-gnueabihf.tar.gz /BurntSushi/ripgrep/releases/download/0.10.0/ripgrep-0.10.0-arm-unknown-linux-gnueabihf.tar.gz /BurntSushi/ripgrep/releases/download/0.9.0/ripgrep-0.9.0-arm-unknown-linux-gnueabihf.tar.gz' ``
> 
> 2. For exa `exa-linux-x86_64-musl-v0.10.1.zip` release was selected over `exa-linux-x86_64-v0.10.1.zip`
>    (might be because of no release containing gnu)
> 
> DEBUG
> .zinit-get-latest-gh-r-url-part DEBUG: `'

@sm1999 Do you mind sharing output of `uname --all`? Can I assume your `$OSTYPE` is `linux-gnu`?

**Comment by sm1999** [comment]
@vladdoster 
uname -all
`Linux SM1999-ARCH 5.16.3-arch1-1 #1 SMP PREEMPT Thu, 27 Jan 2022 14:18:25 +0000 x86_64 GNU/Linux`

I removed all the bpicks. working great for zoxide, gh cli, micro, bat, hyperfine, fd, lsd,
nnn, antidot, glow, fzf, lazygit.

Thank you for your work!

**Comment by vladdoster** [comment]
> @vladdoster uname -all `Linux SM1999-ARCH 5.16.3-arch1-1 #1 SMP PREEMPT Thu, 27 Jan 2022 14:18:25 +0000 x86_64 GNU/Linux`
> 
> I removed all the bpicks. working great for zoxide, gh cli, micro, bat, hyperfine, fd, lsd, nnn, antidot, glow, fzf, lazygit.
> 
> Thank you for your work!

@sm1999 -- Thanks for the info; I'll take a look sometime early this week.

**Comment by sm1999** [comment]
I also found out that `bpick` is not working.

File: zinit-install.zsh [hunk 3]
```diff
@@ -1447,25 +1447,29 @@ builtin source "${ZINIT[BIN_DIR]}/zinit-side.zsh" || {
...
         cygwin  "(windows|cygwin|[-_]win|win64|win32)"
+        darwin  "(((apple|-|_)darwin|mac(os|))^*.AppImage)"
+        linux-musl  "(#i)((musl|linux-musl))"
         windows "(windows|cygwin|[-_]win|win64|win32)"
         msys "(windows|msys|cygwin|[-_]win|win64|win32)"
-        armv7l  "(arm7|armv7)"
-        armv7l-2 "arm7"
-        armv6l  "(arm6|armv6)"
-        armv6l-2 "arm"
+        aarch64-2 "arm"
```
Comment (seagle0128) [inline zinit-install.zsh:1466]: Quick question: why not using `aarch64 "(aarch64|arm)"` ?

**Comment by omares** [comment]
Hey can we somehow help to get this merged? Zinit unfortunately does not work properly on m1 macs.

File: share/git-process-output.zsh [hunk 4]
```diff
@@ -11,16 +11,10 @@ if (( COLS < 10 )) {
...
 typeset -ga progress_frames
 progress_frames=(
-  '0.2 ▹▹▹▹▹ ▸▹▹▹▹ ▹▸▹▹▹ ▹▹▸▹▹ ▹▹▹▸▹ ▹▹▹▹▸'
-  '0.2 ▁ ▃ ▄ ▅ ▆ ▇ ▆ ▅ ▄ ▃'
-  '0.2 ▏ ▎ ▍ ▌ ▋ ▊ ▉ ▊ ▋ ▌ ▍ ▎'
-  '0.2 ▖ ▘ ▝ ▗'
-  '0.2 ◢ ◣ ◤ ◥'
-  '0.2 ▌ ▀ ▐ ▄'
-  '0.2 ✶ ✸ ✹ ✺ ✹ ✷'
+  '0.1 ⠋ ⠙ ⠹ ⠸ ⠼ ⠴ ⠦ ⠧ ⠇ ⠏'
```
Comment (pschmitt) [inline share/git-process-output.zsh:14]: That seems a bit out of scope. But I like it :)

Comment (vladdoster) [reply share/git-process-output.zsh:14]: lol, I forgot I did this. Is out of scope, but I hated the current loading bars. 

I'll open a separate PR for this.

Comment (alichtman) [reply share/git-process-output.zsh:14]: This is really nice though

Comment (vladdoster) [reply share/git-process-output.zsh]: Change made in PR #204

**Comment by seagle0128** [comment]
@vladdoster why was this closed? 😢
